src/common/sec_websocket_extensions.rs

```diff
@@ -0,0 +1,369 @@
+/// value when encoding.
+///
+/// [RFC6455_11.3.2]: https://tools.ietf.org/html/rfc6455#section-11.3.2
+/// [RFC7692_7]: https://tools.ietf.org/html/rfc7692#section-7
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SecWebsocketExtensions(pub Vec<WebsocketExtension>);
+
+impl Header for SecWebsocketExtensions {
+    fn name() -> &'static ::HeaderName {
+        &SEC_WEBSOCKET_EXTENSIONS
+    }
+
+    fn decode<'i, I: Iterator<Item = &'i HeaderValue>>(values: &mut I) -> Result<Self, Error> {
+        let extensions = values
+            .cloned()
+            .flat_map(|v| {
+                FlatCsv::<Comma>::from(v)
+                    .iter()
+                    .map(WebsocketExtension::try_from)
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        if extensions.is_empty() {
+            Err(Error::invalid())
+        } else {
+            Ok(SecWebsocketExtensions(extensions))
+        }
+    }
+
+    fn encode<E: Extend<HeaderValue>>(&self, values: &mut E) {
+        if !self.is_empty() {
+            values.extend(std::iter::once(self.to_value()));
+        }
+    }
+}
+
+impl SecWebsocketExtensions {
+    /// Construct a `SecWebSocketExtensions` from a static string.
+    ///
+    /// ## Panic
+    ///
+    /// Panics if the static string is not a valid extensions valie.
+    pub fn from_static(s: &'static str) -> Self {
+        let value = HeaderValue::from_static(s);
+        SecWebsocketExtensions::try_from(&value).expect("valid static string")
+    }
+
+    /// Convert this `SecWebsocketExtensions` to a single `HeaderValue`.
+    pub fn to_value(&self) -> HeaderValue {
+        let values = self.0.iter().map(HeaderValue::from).collect::<FlatCsv>();
+        HeaderValue::from(&values)
+    }
+
+    /// An iterator over the `WebsocketExtension`s in `SecWebsocketExtensions` header(s).
+    pub fn iter(&self) -> impl Iterator<Item = &WebsocketExtension> {
+        self.0.iter()
+    }
+
+    /// Get the number of extensions.
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns `true` if headers contain no extensions.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+impl TryFrom<&str> for SecWebsocketExtensions {
+    type Error = Error;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        let value = HeaderValue::from_str(value).map_err(|_| Error::invalid())?;
+        SecWebsocketExtensions::try_from(&value)
+    }
+}
+
+impl TryFrom<&HeaderValue> for SecWebsocketExtensions {
+    type Error = Error;
+
+    fn try_from(value: &HeaderValue) -> Result<Self, Self::Error> {
+        let mut values = std::iter::once(value);
+        SecWebsocketExtensions::decode(&mut values)
+    }
+}
+
+/// A WebSocket extension containing the name and parameters.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct WebsocketExtension {
```

Comment:
This just a thought, let me know what you think. What if this type simply held a reference, instead of individual owned value strings and a vector of params? The `name` and `params` methods could return "lazy" values, as a few of the other headers in this crate already do.

Reply:
I actually had `struct SecWebsocketExtensions(HeaderValueString);` with lazy `iter()` originally. But I wanted `decode` to fail on values with invalid syntax, so `typed_try_get` can be used to detect them. This requires parsing the values, so I thought it was better to store the result.

The main reason this can be useful is, when negotiating extensions, values with invalid syntax must be rejected:

> If a value is received by either the client or the server during negotiation that does not conform to the ABNF below, the recipient of such malformed data MUST immediately _Fail the WebSocket Connection_.
>
> https://datatracker.ietf.org/doc/html/rfc6455#section-9.1

If we validate the syntax in `decode`, crates using `headers` can focus on the parameters for each extension, and `warp` can reject the invalid headers, etc..

---

Also, having `WebsocketExtension` independent is useful because the server needs to filter them, and respond with potentially mutated versions (e.g., parameter `("client_max_window_bits", Some("10"))` may be changed to have value `Some("9")` in the response). `WebsocketExtension` can provide setters/builder that makes sure that the value won't result in invalid syntax.

This is why I made the field and the struct public, but I can try to come up with something else if this is an issue.